runtimes/README.md

```diff
@@ -272,6 +272,120 @@ Consider using LSP [ShowMessage notification](https://microsoft.github.io/langua
 | Show notification to customer              | `aws/window/showNotification`     | `NotificationParams`           | [Notification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#notificationMessage) | n/a |
 | Send notification followup back to server   | `aws/window/notificationFollowup`  | `NotificationFollowupParams`    | [Notification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#notificationMessage) | n/a |
 
+### Agent Tools
+In order to create agentic applications Servers can define tools that can be sent to a model backend for use. Tools can be created in multiple servers, and are shared between all servers in the same runtime. There is no JSON/RPC component to Tool usage, Tools are all maintained in Runtime memory.
+
+Tool invocations can use LSP, Workspace or Chat capabilities to interact with the client environment.
+
+The agent feature uses the following interface:
+
+```ts
+export type Agent = {
+    /**
+     * Add a tool to the local tool repository. Tools with the same name will be overwritten.
+     * 
+     * Instead of using the result of this method, for normal tool execution, the tool should be
+     * called using `runTool` to ensure tool name lookup and schema validation happen.
+     * 
+     * @param spec Tool Specification
+     * @param handler The async method to execute when the tool is called
+     * @returns A tool object for local reference.
+     */
+    tool: <T extends InferSchema<S['inputSchema']>, S extends ToolSpec, R>(spec: S, handler: (input: T) => Promise<R>) => Tool<T, R>
+    
+    /**
+     * Run a tool by name. This method will lookup the tool in the local tool repository and
+     * validate the input against the tool's schema.
+     * 
+     * Throws an error if the tool is not found, or if validation fails.
+     *
+     * @param toolName The name of the tool to run
+     * @param input The input to the tool
+     * @returns The result of the tool execution
+     */
+    runTool: (toolName: string, input: any) => Promise<any>
+
+    /**
+     * Get the list of tools in the local tool repository.
+     * @param options Options for the format of the output. Can be either 'bedrock' or 'mcp' (the default)
+     * @returns The tool repository in the requested output format
+     */
+    getTools: <T extends GetToolsOptions>(options?: T) => T extends { format: 'bedrock' } ? BedrockTools : Tools
+}
+```
+
+A typical agent loop would look like this:
+
+```ts
+const AgentServer: Server = ({ chat, agent }) => {
+
+    agent.tool({
+        name: 'use_greeting',
+        description: 'A kind greeting to the user'
+        inputSchema: {
+            type: 'object',
```

Comment:
off-topic: if we are validating schemas at runtime we should consider https://github.com/colinhacks/zod